File: tests/algorithms/transpiler/FS/ciphers/simple_keyword_cypher.fs
// Generated 2025-08-07 10:31 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec index_in_string (s: string) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (String.length (s)) do
            if (string (s.[i])) = ch then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
let rec contains_char (s: string) (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable s = s
    let mutable ch = ch
    try
        __ret <- (index_in_string (s) (ch)) >= 0
        raise Return
        __ret
    with
        | Return -> __ret
let rec is_alpha (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        __ret <- (contains_char (lower) (ch)) || (contains_char (upper) (ch))
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_upper (s: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    try
        let lower: string = "abcdefghijklmnopqrstuvwxyz"
        let upper: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            let ch: string = string (s.[i])
            let idx: int = index_in_string (lower) (ch)
            if idx >= 0 then
                res <- res + (string (upper.[idx]))
            else
                res <- res + ch
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec remove_duplicates (key: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable key = key
    try
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (key)) do
            let ch: string = string (key.[i])
            if (ch = " ") || ((is_alpha (ch)) && ((contains_char (res) (ch)) = false)) then
                res <- res + ch
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec create_cipher_map (key: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable key = key
    try
        let alphabet: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let cleaned: string = remove_duplicates (to_upper (key))
        let mutable cipher: string array = [||]
        let mutable i: int = 0
        while i < (String.length (cleaned)) do
            cipher <- Array.append cipher [|string (cleaned.[i])|]
            i <- i + 1
        let mutable offset: int = String.length (cleaned)
        let mutable j: int = Seq.length (cipher)
        while j < 26 do
            let mutable char: string = string (alphabet.[j - offset])
            while contains_char (cleaned) (char) do
                offset <- offset - 1
                char <- string (alphabet.[j - offset])
            cipher <- Array.append cipher [|char|]
            j <- j + 1
        __ret <- cipher
        raise Return
        __ret
    with
        | Return -> __ret
let rec index_in_list (lst: string array) (ch: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable lst = lst
    let mutable ch = ch
    try
        let mutable i: int = 0
        while i < (Seq.length (lst)) do
            if (_idx lst (i)) = ch then
                __ret <- i
                raise Return
            i <- i + 1
        __ret <- -1
        raise Return
        __ret
    with
        | Return -> __ret
let rec encipher (message: string) (cipher: string array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable message = message
    let mutable cipher = cipher
    try
        let alphabet: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let msg: string = to_upper (message)
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (msg)) do
            let ch: string = string (msg.[i])
            let idx: int = index_in_string (alphabet) (ch)
            if idx >= 0 then
                res <- res + (_idx cipher (idx))
            else
                res <- res + ch
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec decipher (message: string) (cipher: string array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable message = message
    let mutable cipher = cipher
    try
        let alphabet: string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let msg: string = to_upper (message)
        let mutable res: string = ""
        let mutable i: int = 0
        while i < (String.length (msg)) do
            let ch: string = string (msg.[i])
            let idx: int = index_in_list (cipher) (ch)
            if idx >= 0 then
                res <- res + (string (alphabet.[idx]))
            else
                res <- res + ch
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let cipher_map: string array = create_cipher_map ("Goodbye!!")
let encoded: string = encipher ("Hello World!!") (cipher_map)
printfn "%s" (encoded)
printfn "%s" (decipher (encoded) (cipher_map))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/Java/ciphers/simple_keyword_cypher.java
public class Main {
    static String[] cipher_map;
    static String encoded;

    static int index_in_string(String s, String ch) {
        int i = 0;
        while (i < _runeLen(s)) {
            if ((s.substring(i, i+1).equals(ch))) {
                return i;
            }
            i = i + 1;
        }
        return -1;
    }

    static boolean contains_char(String s, String ch) {
        return index_in_string(s, ch) >= 0;
    }

    static boolean is_alpha(String ch) {
        String lower = "abcdefghijklmnopqrstuvwxyz";
        String upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        return ((Boolean)(contains_char(lower, ch))) || ((Boolean)(contains_char(upper, ch)));
    }

    static String to_upper(String s) {
        String lower_1 = "abcdefghijklmnopqrstuvwxyz";
        String upper_1 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String res = "";
        int i_1 = 0;
        while (i_1 < _runeLen(s)) {
            String ch = s.substring(i_1, i_1+1);
            int idx = index_in_string(lower_1, ch);
            if (idx >= 0) {
                res = res + upper_1.substring(idx, idx+1);
            } else {
                res = res + ch;
            }
            i_1 = i_1 + 1;
        }
        return res;
    }

    static String remove_duplicates(String key) {
        String res_1 = "";
        int i_2 = 0;
        while (i_2 < _runeLen(key)) {
            String ch_1 = key.substring(i_2, i_2+1);
            if ((ch_1.equals(" ")) || (((Boolean)(is_alpha(ch_1))) && contains_char(res_1, ch_1) == false)) {
                res_1 = res_1 + ch_1;
            }
            i_2 = i_2 + 1;
        }
        return res_1;
    }

    static String[] create_cipher_map(String key) {
        String alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String cleaned = String.valueOf(remove_duplicates(String.valueOf(to_upper(key))));
        String[] cipher = ((String[])(new String[]{}));
        int i_3 = 0;
        while (i_3 < _runeLen(cleaned)) {
            cipher = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(cipher), java.util.stream.Stream.of(cleaned.substring(i_3, i_3+1))).toArray(String[]::new)));
            i_3 = i_3 + 1;
        }
        int offset = _runeLen(cleaned);
        int j = cipher.length;
        while (j < 26) {
            String char_ = alphabet.substring(j - offset, j - offset+1);
            while (contains_char(cleaned, char_)) {
                offset = offset - 1;
                char_ = alphabet.substring(j - offset, j - offset+1);
            }
            cipher = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(cipher), java.util.stream.Stream.of(char_)).toArray(String[]::new)));
            j = j + 1;
        }
        return cipher;
    }

    static int index_in_list(String[] lst, String ch) {
        int i_4 = 0;
        while (i_4 < lst.length) {
            if ((lst[i_4].equals(ch))) {
                return i_4;
            }
            i_4 = i_4 + 1;
        }
        return -1;
    }

    static String encipher(String message, String[] cipher) {
        String alphabet_1 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String msg = String.valueOf(to_upper(message));
        String res_2 = "";
        int i_5 = 0;
        while (i_5 < _runeLen(msg)) {
            String ch_2 = msg.substring(i_5, i_5+1);
            int idx_1 = index_in_string(alphabet_1, ch_2);
            if (idx_1 >= 0) {
                res_2 = res_2 + cipher[idx_1];
            } else {
                res_2 = res_2 + ch_2;
            }
            i_5 = i_5 + 1;
        }
        return res_2;
    }

    static String decipher(String message, String[] cipher) {
        String alphabet_2 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String msg_1 = String.valueOf(to_upper(message));
        String res_3 = "";
        int i_6 = 0;
        while (i_6 < _runeLen(msg_1)) {
            String ch_3 = msg_1.substring(i_6, i_6+1);
            int idx_2 = index_in_list(((String[])(cipher)), ch_3);
            if (idx_2 >= 0) {
                res_3 = res_3 + alphabet_2.substring(idx_2, idx_2+1);
            } else {
                res_3 = res_3 + ch_3;
            }
            i_6 = i_6 + 1;
        }
        return res_3;
    }
    public static void main(String[] args) {
        cipher_map = ((String[])(create_cipher_map("Goodbye!!")));
        encoded = String.valueOf(encipher("Hello World!!", ((String[])(cipher_map))));
        System.out.println(encoded);
        System.out.println(decipher(encoded, ((String[])(cipher_map))));
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }
}
